ring.go

```diff
@@ -275,14 +275,14 @@ func (c *ringShards) Heartbeat(frequency time.Duration) {
 func (c *ringShards) rebalance() {
 	hash := newConsistentHash(c.opt)
 	var shardsNum int
+	c.mu.Lock()
```

Comment:
Is this necessary? `hash` is a local var and `shard.IsUp` uses atomics to it should be safe to be used in multiple goroutines.

Reply:
the problem is on `for name, shard := range c.shards`, you read from `c.shards` while `Ring.Close` can set it to `nil`, so the race condition (you can check travis build fails for the first commit)

Reply:
maybe I can only `RLock` around the range and then keep the full lock before writing to `c.hash`